Add husqvarna automower ble integration

alistair23 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This PR adds support for the Husqvarna Automower BLE integration.

The Husqvarna Automower BLE integration provides connectivity with Husqvarna Automowers lawn mowers via a local Bluetooth connection. This allows connecting and controlling an Automower without any accounts, cloud or network connection.

The integration is based on [AutoMower-BLE](https://github.com/alistair23/AutoMower-BLE), a unofficial reverse engineered Husqvarna Automower Connect BLE library.

The integration has been tested against a 305 Automower using a ESPHome Bluetooth proxy.

This is somewhat similar to https://github.com/home-assistant/core/pull/100569, except https://github.com/home-assistant/core/pull/100569 uses the cloud API while this integration uses the local BLE. Not all mowers support BLE and not all mowers support the cloud, so although there is overlap both integrations support different mowers.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/30921

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

alistair23 committed:
husqvarna_automower_ble: Initial commit

Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented:
Never really worked a lot with bluetooth so I can't help you on that part

Thomas55555 commented on tests/components/cast/test_media_player.py:
```diff
@@ -2285,7 +2285,6 @@ async def test_cast_platform_play_media_local_media(
-        app_data["media_id"]
-        == f"{network.get_url(hass)}/api/hls/bla/master_playlist.m3u8?token=bla"
-    )
+    assert app_data[
+        "media_id"
+    ] == f"{network.get_url(hass)}/api/hls/bla/master_playlist.m3u8?token=bla"
```
Not related

Thomas55555 commented:
I'm also not, and I'm self starting to learn here. I just gave you some general review.

alistair23 commented:
I have updated this to support HA `2024.4.0`. For everyone who is manually pulling this component in, don't update the component until you are running HA 2024.4 otherwise it won't work. If you have updated to 2024.4 you will need to update this component as well.

The tests are now running in the CI as well, which is great. The current `Split tests for full run` failure is unrelated to this PR

alistair23 commented:
If you want to try this out while waiting for it to be reviewed all of the documentation, including steps on installing this PR before it's merged, are available at: https://github.com/home-assistant/home-assistant.io/pull/30921#issuecomment-2017744301

GitWally commented:
Tested with Husqvarna 305 and ESP32 BT Proxy on HA Supervised: in my case it works, and finally I can control my lawn mower!

emontnemery reviewed (changes requested):
There's already a husqvarna_automower integration, and we need a Husqvarna [brand](https://developers.home-assistant.io/docs/creating_integration_brand?_highlight=brand) to help the user pick the integration which is right for them.

Please add that to this PR.

jenssvedlund commented:
Great progress, many thanks for your work! 
Just tried it with my Automower 310 MARK II, but failing after it retrieves the initial data.
How do you want manage issues regarding the project?

alistair23 commented:
Brand information has been added

emontnemery reviewed (dismissed):
Tests are failing:
```
FAILED tests/components/husqvarna_automower_ble/test_config_flow.py::test_user_selection - KeyError: 'path'
FAILED tests/components/husqvarna_automower_ble/test_config_flow.py::test_bluetooth - KeyError: 'path'
FAILED tests/components/husqvarna_automower_ble/test_config_flow.py::test_bluetooth_invalid - KeyError: 'path'
FAILED tests/components/husqvarna_automower_ble/test_init.py::test_setup - AssertionError: assert [+ received] == [- snapshot]
  Snapshot 'test_setup' does not exist!
  + None
ERROR tests/components/husqvarna_automower_ble/test_config_flow.py::test_user_selection - KeyError: 'path'
ERROR tests/components/husqvarna_automower_ble/test_config_flow.py::test_no_devices - KeyError: 'path'
```

Please fix the tests, then mark the PR as ready for review 👍

mkmer commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,190 @@
+        self.async_write_ha_state()
+
+
+class BatterySensor(HusqvarnaAutomowerBleEntity, SensorEntity):
```
Sensor entities should be in a sensor platform file `sensor.py`

mkmer commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -0,0 +1,67 @@
+LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = [
+    Platform.LAWN_MOWER,
```
Should also include `Platform.SENSOR` and separate sensor entities to the sensor platform

mkmer commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,190 @@
+                | LawnMowerEntityFeature.START_MOWING
+                | LawnMowerEntityFeature.DOCK,
+            ),
+            BatterySensor(
+                coordinator,
+                "automower" + model + "_" + address + "_battery_level",
+                "Battery Level",
+            ),
+        ]
+    )
```
Sensor should be in sensor platform.

mkmer commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -0,0 +1,115 @@
+        self._abort_if_unique_id_configured()
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(
```
Something flaky goes on when setting up and paring the device the first time.
1)  I paired and the setup "hung" at the pairing state.
2) Reloaded the integration and it found entities but they were unavailable.
3) Reloaded again and the entities became available.

On a second try with the esphome ble gateway already "paired" I still needed to do a reload to get the entities to become available when adding the integration.

alistair23 replied:
Can you provide logs?

I found pairing the official application unreliable for the first pair as well

florian-d replied:
> I found pairing the official application unreliable for the first pair as well

Same here

mkmer commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -0,0 +1,67 @@
+        hass, address, connectable=True
+    ) or await get_device(address)
+    if not await mower.connect(device):
+        return False
```
I don't think we want to "fail" the integration here.  We have already "validated" the connectivity in the config flow.  At this point, it's just not in range and should continue to try.  Simply `raise  ConfigEntryNotReady()` to let HA retry automatically.

mkmer commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,121 @@
+            _LOGGER.error("Can't find device")
+            raise UpdateFailed("Can't find device")
+
+        if not await self.mower.connect(device):
```
We should `try:` to connect to the mower and catch `except: bleak.exc.BleakError)` : Error ESP_GATT_CONN_FAIL_ESTABLISH while connecting: Connection failed to establish.

mkmer commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,121 @@
+
+        data: dict[str, bytes] = {}
+
+        if not self.mower.is_connected():
```
`try:` around this section as well

mkmer commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -0,0 +1,67 @@
+    await close_stale_connections_by_address(address)
+
+    LOGGER.debug("connecting to %s with channel ID %s", address, str(channel_id))
+    device = bluetooth.async_ble_device_from_address(
```
Bleak errors should be caught here.

mkmer commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,190 @@
+"""Demo platform that has a couple fake lawn mowers."""
```
Stale doc string?

mkmer commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,121 @@
+SCAN_INTERVAL = timedelta(seconds=60)
+
+
+class Coordinator(DataUpdateCoordinator[dict[str, bytes]]):
```
Should name this something like `HusqvarnaCoordinator` to help prevent confusion with the normal `coordinator` - like other integrations do.

mkmer commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,127 @@
+        super().__init__(
+            hass=hass,
+            logger=logger,
+            name="Husqvarna Automower BLE Data Update Coordinator",
```
Based on other examples, this is usually = DOMAIN (not sure if it really matters)

mkmer commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -0,0 +1,77 @@
+    model = await mower.get_model()
+    LOGGER.info("Connected to Automower: %s", model)
+
+    device_info = DeviceInfo(
```
Each entity should have it's own DeviceInfo to put in the device registry.  I think it needs to move the the entity setup for each.

alistair23 commented:
@emontnemery are you able to help shed some light on the test failure? I can not figure out why it's failing

alistair23 commented:
This is the error that I'm seeing that I can not figure out

```
File "/home/runner/work/core/core/tests/components/husqvarna_automower_ble/test_config_flow.py", line 28, in test_user_selection
    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
  File "/home/runner/work/core/core/tests/components/bluetooth/__init__.py", line 205, in inject_bluetooth_service_info
    inject_advertisement(hass, device, advertisement_data)
  File "/home/runner/work/core/core/tests/components/bluetooth/__init__.py", line 111, in inject_advertisement
    return inject_advertisement_with_source(hass, device, adv, SOURCE_LOCAL)
  File "/home/runner/work/core/core/tests/components/bluetooth/__init__.py", line 118, in inject_advertisement_with_source
    inject_advertisement_with_time_and_source(
  File "/home/runner/work/core/core/tests/components/bluetooth/__init__.py", line 131, in inject_advertisement_with_time_and_source
    inject_advertisement_with_time_and_source_connectable(
  File "/home/runner/work/core/core/tests/components/bluetooth/__init__.py", line 145, in inject_advertisement_with_time_and_source_connectable
    async_get_advertisement_callback(hass)(
  File "/home/runner/work/core/core/homeassistant/components/bluetooth/manager.py", line 114, in _discover_service_info
    discovery_flow.async_create_flow(
  File "/home/runner/work/core/core/homeassistant/helpers/discovery_flow.py", line 36, in async_create_flow
    hass.async_create_background_task(
  File "/home/runner/work/core/core/homeassistant/core.py", line 851, in async_create_background_task
    task = create_eager_task(target, name=name, loop=self.loop)
  File "/home/runner/work/core/core/homeassistant/util/async_.py", line 37, in create_eager_task
    return Task(coro, loop=loop, name=name, eager_start=True)
Traceback (most recent call last):
  File "/home/runner/work/core/core/homeassistant/loader.py", line 1121, in async_get_platforms
    await self.hass.async_add_import_executor_job(
asyncio.exceptions.CancelledError: Home Assistant is stopping

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/runner/work/core/core/homeassistant/config_entries.py", line 1250, in async_init
    flow, result = await self._async_init(flow_id, handler, context, data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/core/core/homeassistant/config_entries.py", line 1270, in _async_init
    flow = await self.async_create_flow(handler, context=context, data=data)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/core/core/homeassistant/config_entries.py", line 1406, in async_create_flow
    handler = await _async_get_flow_handler(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/core/core/homeassistant/config_entries.py", line 2764, in _async_get_flow_handler
    await _load_integration(hass, domain, hass_config)
  File "/home/runner/work/core/core/homeassistant/config_entries.py", line 2743, in _load_integration
    await integration.async_get_platform("config_flow")
  File "/home/runner/work/core/core/homeassistant/loader.py", line 1074, in async_get_platform
    platforms = await self.async_get_platforms((platform_name,))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/core/core/homeassistant/loader.py", line 1146, in async_get_platforms
    import_future.set_exception(ex)
asyncio.exceptions.InvalidStateError: invalid state
```

alistair23 commented:
Tests are passing and comments are addressed

wtaferner commented:
Happy to see this PR and important feature to be added. Thx @alistair23 for the quite huge effort. 

@emontnemery 
Any chance that it might make it to the June release?

mkmer commented:
> > comments are addressed
> 
> What about this comment though: [#108326 (comment)](https://github.com/home-assistant/core/pull/108326#discussion_r1599248788), users should not have to reload the integration to make it work. What happens if the automower is out of bluetooth range?

I don't think it's possible to configure the integration if the automower is out of range (it won't / can't be discovered).  
I believe we fixed the issue when it's already configured and out of range at startup, it automatically retries by raising `ConfigEntryNotReady`.

joostlek commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,126 @@
+        return data
+
+
+class HusqvarnaAutomowerBleEntity(CoordinatorEntity[HusqvarnaCoordinator]):
```
Please move this to `entity.py`

joostlek commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,126 @@
+    @property
+    def available(self) -> bool:
+        """Return if entity is available."""
+        return self.coordinator.mower.is_connected()
```
We usually also incorporate the super().available in coordinator entities

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+        coordinator: HusqvarnaCoordinator,
+        unique_id: str,
+        name: str,
+        features: LawnMowerEntityFeature = LawnMowerEntityFeature(0),
```
Why is this a parameter if it's static?

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+    def __init__(
+        self,
+        coordinator: HusqvarnaCoordinator,
+        unique_id: str,
+        name: str,
```
Why don't we just pass in what we need and let the entity decide

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+    ) -> None:
+        """Initialize the lawn mower."""
+        super().__init__(coordinator)
+        self._attr_name = name
```
The lawn mower is the main feature of the device, so the name of the lawn mower should be set to `_attr_name = None`. This way, the name of this entity is decided by the name of the device, which is consistent

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+        [
+            AutomowerLawnMower(
+                coordinator,
+                "automower" + str(model) + "_" + str(address),
```
Also, for unique_id, address should be purely unique I assume? Reminder that unique_id is unique per platform per integration, so there's no need to prepend it with automower (if that's what you were trying to avoid)

alistair23 replied:
Yeah, address should be unique. I'll drop the other bits

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+        state = str(self.coordinator.data["state"])
+        activity = str(self.coordinator.data["activity"])
+
+        if state is None:
+            return None
+
+        if activity is None:
+            return None
```
can be combined

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,156 @@
+                return
+
+        await self.coordinator.mower.mower_resume()
+        if self._attr_activity == LawnMowerActivity.DOCKED:
```
```suggestion
        if self._attr_activity is LawnMowerActivity.DOCKED:
```

mkmer commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+        channel_id = random.randint(1, 0xFFFFFFFF)
+
+        try:
+            (manufacture, device_type, model) = await Mower(
```
This call appears to return a `bool`
```                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower_ble/config_flow.py", line 117, in async_step_user
    return await self.async_step_confirm()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/husqvarna_automower_ble/config_flow.py", line 81, in async_step_confirm
    (manufacture, device_type, model) = await Mower(
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: cannot unpack non-iterable bool object```

al31c0 commented:
I have been following this integration for a long time. Is it possible to have it in the release HA 2024.7? 
@elupus @emontnemery

wtaferner commented:
@al31c0 
I guess there are some pending adaptations needed by @alistair23, but he seems busy or at least nothing moved lately here. I am also looking forward to seeing this one merged, but I guess as long as you do not contribute or be able to move it forward yourself, we can just cheer, be thankful and patient for things to happen at some point 🙏🏼

alistair23 commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,162 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result == snapshot
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={"address": "00000000-0000-0000-0000-000000000001"},
+    )
+    assert result == snapshot
```
Is there a more preferred way to do this?

joostlek replied:
Oh lol, I can't even remember I reviewed this 2 months ago.

Well, just having a few asserts

alistair23 commented:
I have addressed all of the comments except for the snapshot comment. I'm not sure what I should use instead.

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,147 @@
+        if self.coordinator.data is None:
+            return None
+
+        state = str(self.coordinator.data["state"])
+        activity = str(self.coordinator.data["activity"])
```
Would it be an idea to do this in the coordinator?

joostlek replied:
I think you could use a dataclass to replace the `dict[str, bytes]` and make it more typesafe

alistair23 replied:
The Python library has been re-written to support [enums for these](https://github.com/alistair23/AutoMower-BLE/blob/main/automower_ble/protocol.py#L24)

Once this PR is merged the next step will be to update to the not yet released re-write. At that point this can be updated if that's ok. I'm trying to get this merged first before handling the API changes with a large library update

joostlek replied:
Why don't we use the new lib already?

alistair23 replied:
Basically because this current PR is using the existing one and I want to get this merged :) I wanted to avoid changing this PR to try and keep things simple

Then there will be follow up work to update to the new lib and support more request types

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,147 @@
+
+        self._attr_activity = self._get_activity()
+        self._attr_available = self._attr_activity is not None
+        self.async_write_ha_state()
```
```suggestion
        super()._handle_coordinator_update()
```

alistair23 commented on homeassistant/components/husqvarna_automower_ble/coordinator.py:
```diff
@@ -0,0 +1,100 @@
+        try:
+            data["battery_level"] = await self.mower.battery_level()
+            LOGGER.debug(data["battery_level"])
+            if data["battery_level"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
+
+            data["activity"] = await self.mower.mower_activity()
+            LOGGER.debug(data["activity"])
+            if data["activity"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
+
+            data["state"] = await self.mower.mower_state()
+            LOGGER.debug(data["state"])
+            if data["state"] is None:
+                await self._async_find_device()
+                raise UpdateFailed("Error getting data from device")
```
A update involves sending a request to the mower for every type of data (currently battery, state and activity) and waiting for a response. If the connection is lost (because the mower moves out of range or stops responding) then it's possible to only get some of the data. In reality it's not uncommon to happen.

So this is trying to recover the device if any of the data queries fail. If the mower if back in range then that's great, which is what we are trying to catch.

The other issue is that the mower can sometimes just stop responding for a little bit, so we are also trying to work around that.

joostlek replied:
Would it be viable to merge them all? Because they all raise UpdateFailed, and if that happens, all coordinator entities are marked unavailable

alistair23 replied:
The idea is to error as soon as we can't get a single query. If we consolidate them all then we have to wait until the end to retry connecting. Which just uses more time

joostlek replied:
Right

joostlek commented on homeassistant/components/husqvarna_automower_ble/lawn_mower.py:
```diff
@@ -0,0 +1,147 @@
+    ) -> None:
+        """Initialize the lawn mower."""
+        super().__init__(coordinator)
+        self._attr_name = None
```
I mean, this does the same, but since it's static you can just move it outside of the constructor

joostlek replied:
I mean, this does the same, but since it's static you can just move it outside of the constructor

joostlek commented on homeassistant/components/husqvarna_automower_ble/__init__.py:
```diff
@@ -0,0 +1,82 @@
+]
+
+
+@dataclass
+class HusqvarnaAutomowerBleData:
+    """Data for AccuWeather integration."""
+
+    coordinator: HusqvarnaCoordinator
+
+
+type HusqvarnaAutomowerBleConfigEntry = ConfigEntry[HusqvarnaAutomowerBleData]
```
If we're only storing the coordinator, let's not wrap it in a dataclass yet

joostlek commented on homeassistant/components/husqvarna_automower_ble/entity.py:
```diff
@@ -0,0 +1,34 @@
+
+        device_info = DeviceInfo(
+            identifiers={
+                (DOMAIN, str(coordinator.address) + str(coordinator.channel_id))
```
```suggestion
                (DOMAIN, f"{coordinator.address}_{coordinator.channel_id}")
```

joostlek commented on tests/components/gardena_bluetooth/test_config_flow.py:
```diff
@@ -31,6 +31,7 @@ async def test_user_selection(
 
     inject_bluetooth_service_info(hass, WATER_TIMER_SERVICE_INFO)
     inject_bluetooth_service_info(hass, WATER_TIMER_UNNAMED_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
```
Why do we change Gardena tests?

joostlek commented on tests/components/husqvarna_automower_ble/conftest.py:
```diff
@@ -0,0 +1,114 @@
+
+
+@pytest.fixture
+def mock_entry():
```
Please type

joostlek commented on tests/components/husqvarna_automower_ble/conftest.py:
```diff
@@ -0,0 +1,114 @@
+@pytest.fixture
+def mock_entry():
+    """Create hass config fixture."""
+    return MockConfigEntry(
```
I think we have an unique id in real life, so let's make this mock config entry reflect real life

alistair23 replied:
Unless I'm misunderstanding I don't think we do

joostlek commented on tests/components/husqvarna_automower_ble/conftest.py:
```diff
@@ -0,0 +1,114 @@
+
+    freezer.move_to("2023-01-01T01:00:00Z")
+
+    async def delay():
```
Typing

mkmer commented:
Really should stop force pushing commits - it makes following up on reviews difficult.

joostlek commented on tests/components/husqvarna_automower_ble/conftest.py:
```diff
@@ -0,0 +1,114 @@
+class MockMower:
+    """Mock BleakClient."""
+
+    def __init__(self, *args, **kwargs):
+        """Mock BleakClient."""
+
+    async def __aexit__(self, *args, **kwargs):
+        """Mock BleakClient.__aexit__."""
+
+    async def connect(self, *args, **kwargs) -> bool:
+        """Mock BleakClient.connect."""
+        return True
+
+    async def disconnect(self, *args, **kwargs):
+        """Mock BleakClient.disconnect."""
+
+    def is_connected(self):
+        """Mock BleakClient.is_connected."""
+        return True
+
+    async def get_model(self) -> str:
+        """Mock BleakClient.get_model."""
+        return "305"
+
+    async def battery_level(self) -> int:
+        """Mock BleakClient.battery_level."""
+        return 100
+
+    async def mower_state(self) -> str:
+        """Mock BleakClient.mower_state."""
+        return "pendingStart"
+
+    async def mower_activity(self) -> str:
+        """Mock BleakClient.mower_activity."""
+        return "charging"
+
+    async def probe_gatts(self, device):
+        """Mock BleakClient.probe_gatts."""
+        return ("Husqvarna", "Automower", "305")
+
```
Honestly, use mocks. AsyncMocks are very useful in cases like this. I once went this approach with `twitch` and I regretted it big time because I had to change logic like raising exceptions, and it was becoming a lot of ducttaping. 

Instead I personally really like asyncmocks, checkout AirGradient tests for an example. They're really flexible and easy to maintain

alistair23 replied:
I don't really follow. I am using Mocks and it doesn't look that different to the AirGradient tests

joostlek replied:
You're creating a MockMower object. We should just use an AsyncMock instead

joostlek commented on tests/components/husqvarna_automower_ble/conftest.py:
```diff
@@ -0,0 +1,114 @@
+    """Auto mock bluetooth."""
+
+    with (
+        patch(
+            "automower_ble.mower.Mower",
+            MockMower,
+        ),
```
Only mock where we use the library

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,163 @@
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,163 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result == snapshot
```
Like what I mean with my other comment about snapshot testing, I am reading this test now and I am like, wait, what are we testing here? And I have to scroll back and check the test name to know what we're testing. But then I still don't know what this _actually_ tests. Does it show an error form? Does it abort? Nobody knows, except for the ones brave enough to open the ambr file

alistair23 replied:
Fair point! I get it now, I have switched this to asserts

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,163 @@
+    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    result = next(iter(hass.config_entries.flow.async_progress_by_handler(DOMAIN)))
```
Can't we just do `[0]`

alistair23 replied:
Instead of the `next()` I don't think so

joostlek replied:
I think instead of everything

joostlek commented on tests/components/husqvarna_automower_ble/test_init.py:
```diff
@@ -0,0 +1,392 @@
+    assert device == snapshot
+
+    coordinator = mock_entry.runtime_data.coordinator
+    await coordinator.async_shutdown()
```
Don't touch internals. Rather check if we really did the things needed for a shutdown at the (soon to be added) mocks

alistair23 committed:
husqvarna_automower_ble: Remove coordinator dataclass

Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented on tests/components/husqvarna_automower_ble/test_init.py:
```diff
@@ -0,0 +1,392 @@
+    device_registry: dr.DeviceRegistry,
+    mock_client: Mock,
+    mock_entry: MockConfigEntry,
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test setup creates expected devices."""
+
+    async def _connect(self, *args, **kwargs) -> bool:
+        """Mock BleakClient.connect."""
+        return False
+
+    mock_client.connect = _connect
+
+    mock_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert mock_entry.state is ConfigEntryState.SETUP_RETRY
+
+
+async def test_setup_failed_connect(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    mock_client: Mock,
+    mock_entry: MockConfigEntry,
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test setup creates expected devices."""
+
+    async def _connect(self, *args, **kwargs) -> bool:
+        """Mock BleakClient.connect."""
+        raise BleakError
+
+    mock_client.connect = _connect
+
+    mock_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_entry.entry_id)
```
You probably can use parametrize for this and you'll love mocks here

joostlek commented on tests/components/husqvarna_automower_ble/test_init.py:
```diff
@@ -0,0 +1,392 @@
+
+    mock_client.is_connected = _is_connected
+
+    await coordinator._async_update_data()
```
Don't touch internals

joostlek commented on tests/components/husqvarna_automower_ble/test_init.py:
```diff
@@ -0,0 +1,392 @@
+
+    mock_client.battery_level = og_battery_level
+
+
+async def test_setup_exception_battery(
+    hass: HomeAssistant,
+    device_registry: dr.DeviceRegistry,
+    mock_client: Mock,
+    mock_entry: MockConfigEntry,
+    snapshot: SnapshotAssertion,
+) -> None:
+    """Test setup creates expected devices."""
+
+    def _is_connected_success(self, *args, **kwargs) -> bool:
+        """Mock BleakClient.is_connected."""
+        return True
+
+    async def _connect_success(self, *args, **kwargs) -> bool:
+        """Mock BleakClient.connect."""
+        return True
+
+    mock_client.connect = _connect_success
+    mock_client.is_connected = _is_connected_success
+
+    mock_entry.add_to_hass(hass)
+    await hass.config_entries.async_setup(mock_entry.entry_id)
+    await hass.async_block_till_done()
+
+    assert mock_entry.state is ConfigEntryState.LOADED
+
+    async def _battery_level(self, *args, **kwargs):
+        raise BleakError
+
+    og_battery_level = mock_client.battery_level
+    mock_client.battery_level = _battery_level
+
+    coordinator = mock_entry.runtime_data
+
+    with contextlib.suppress(UpdateFailed):
+        await coordinator._async_update_data()
```
I think these can be combined via parametrize and mocks

joostlek reviewed (changes requested):
Feel free to send me a message on discord. I think we can make this in for the next release

joostlek commented:
Oh yes, the force push is annoying me a bit too as I can't easy check what changed

alistair23 committed:
tests: husqvarna_automower_ble: Remove snapshots

Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,173 @@
+        result["flow_id"],
+        user_input={},
+    )
+    assert result["title"] == "Husqvarna Automower"
+    assert result["handler"] == "husqvarna_automower_ble"
```
Oh let me be more concrete: let's test result type, entry data, entry unique id and title

joostlek commented on homeassistant/components/husqvarna_automower_ble/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+            return self.async_abort(reason="no_devices_found")
+
+        self.address = discovery_info.address
+        await self.async_set_unique_id(self.address)
+        self._abort_if_unique_id_configured()
```
Re: to other comment, you do :)

alistair23 committed:
tests: husqvarna_automower_ble: Specify CONF_UNIQUE_ID

Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,176 @@
+    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    result = next(iter(hass.config_entries.flow.async_progress_by_handler(DOMAIN)))
```
```suggestion
    result = hass.config_entries.flow.async_progress_by_handler(DOMAIN)[0]
```

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,176 @@
+    assert result["handler"] == "husqvarna_automower_ble"
+
+    assert result["data"]["address"] == "00000000-0000-0000-0000-000000000003"
+    assert result["context"]["unique_id"] == "00000000-0000-0000-0000-000000000003"
```
```suggestion
    assert result["result"]["unique_id"] == "00000000-0000-0000-0000-000000000003"
```

alistair23 replied:
```python
        result = hass.config_entries.flow.async_progress_by_handler(DOMAIN)[0]
        assert result["step_id"] == "confirm"
>       assert result["result"]["unique_id"] == "00000000-0000-0000-0000-000000000003"
E       KeyError: 'result'

tests/components/husqvarna_automower_ble/test_config_flow.py:84: KeyError
```

joostlek replied:
That is strange, I think we need to make it so that it finalizes (aka, flowresulttype.CREATE_ENTRY

joostlek commented:
For the config flow asserts, please check other integrations, I think they can be a bit better because we don't need to assert the handler for example

joostlek commented:
I am now also looking at the code to give it a final smash. Why is the device identifier address+channel? I feel like I have asked this before but couldn't see quickly in the comments I made

joostlek commented:
Can you please read the comments? :)

alistair23 commented:
> Can you please read the comments? :)

Yep, just rebasing before I try to re-write the tests

joostlek commented:
Mostly the, please don't force push part and feel free to send me a message on discord

mkmer commented on homeassistant/components/time/__init__.py:
Unrelated changes

mkmer reviewed (commented):
So close - Maybe we can make it in the September release...

alistair23 commented:
I'm adding changes on top as separate commits, but I need to force push the rebases on `dev`.

I'm currently trying to figure out how to re-write the tests in a `AsyncMock` way, but it's slow going.

I'm not on Discord

joostlek commented:
Check! Let me know if there is anything I can help you with, would be awesome to get it in 2024.10

alistair23 commented:
Maybe any hints on how to use AsyncMock? I'm looking at the Air Gradient code, but I keep getting errors like

```
RuntimeError: BluetoothManager has not been set
```

and

```
ConfigEntryState.NOT_LOADED
```

joostlek commented:
Oh maybe this needs an extra fixture to completely mock out the Bluetooth. Maybe check at other bluetooth integrations on what they do. When I'm back at my pc in a few minutes I'll also search for you

joostlek commented:
Oh you already have that one hmmm

joostlek commented:
Is the version here the latest version? I'll check it out locally and can give it a stab

alistair23 commented:
Thanks!

This isn't the latest, but I think I figured it out. I'll push again in a sec

alistair23 committed:
tests: husqvarna_automower_ble: Convert to AsyncMock

Signed-off-by: Alistair Francis <alistair@alistair23.me>

joostlek commented:
The test run good for me locally

alistair23 commented:
Yep, I think the tests are working.

Now trying to figure out: https://github.com/home-assistant/core/pull/108326#discussion_r1687970286

joostlek commented:
I made some cleanups, mind if I push?

alistair23 commented:
Go for it

joostlek committed:
Clean up, add strings

joostlek commented:
Apparently you were also missing a `strings.json` file. So we should check if the one I added now is complete, and it needs an explanation for the no_device_found error and add basic troubleshooting steps (like `switchbot` has)

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,192 @@
+    }
+
+
+async def test_no_devices(hass: HomeAssistant) -> None:
+    """Test missing device."""
+
+    inject_bluetooth_service_info(hass, AUTOMOWER_MISSING_SERVICE_SERVICE_INFO)
+    inject_bluetooth_service_info(hass, AUTOMOWER_UNSUPPORTED_GROUP_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_USER}
+    )
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == "user"
```
What should this test test?

alistair23 replied:
In theory it's testing for a device with a missing `service_uuids`

I think I just copied this from the Gardena tests, and modified it for the AutoMower, so it might not be very relevant. But I thought it seemed useful to keep

joostlek replied:
Yea but it's essentially testing nothing, it injects 2 services and then starts a user flow but it doesn't really do anything

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,192 @@
+    }
+
+
+async def test_exception_connect(
+    hass: HomeAssistant,
+    mock_automower_client: Mock,
+) -> None:
+    """Test we can select a device."""
+
+    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
+    inject_bluetooth_service_info(hass, AUTOMOWER_UNNAMED_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    mock_automower_client.probe_gatts.side_effect = BleakError
+
+    result = hass.config_entries.flow.async_progress_by_handler(DOMAIN)[0]
+    assert result["step_id"] == "confirm"
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={},
+    )
+    assert result["type"] is FlowResultType.ABORT
+    assert result["reason"] == "cannot_connect"
```
I modified this test to abort when trying to confirm it

joostlek commented on tests/components/husqvarna_automower_ble/test_config_flow.py:
```diff
@@ -0,0 +1,192 @@
+    assert result["reason"] == "cannot_connect"
+
+
+async def test_failed_is_connected(
+    hass: HomeAssistant,
+    mock_automower_client: Mock,
+) -> None:
+    """Test we can select a device."""
+
+    inject_bluetooth_service_info(hass, AUTOMOWER_SERVICE_INFO)
+    inject_bluetooth_service_info(hass, AUTOMOWER_UNNAMED_SERVICE_INFO)
+    await hass.async_block_till_done(wait_background_tasks=True)
+
+    mock_automower_client.is_connected.side_effect = False
```
What should this test test?

alistair23 replied:
If the device disconnects during the config_flow process

joostlek replied:
But what should happen then? Should we check if the flow that is started is aborted?

alistair23 replied:
Yeah, I think so. The mower BLE connection can be a little unreliable. So we probably just want to abort and have the user try again

joostlek replied:
Yes but the user isn't trying anything, so I'm wondering what would happen at this point

joostlek reviewed (commented):
Some things I found while checking out

joostlek commented:
If you can check out these 3 things, I'll take a stab on the init

joostlek committed:
Fix init tests

joostlek commented:
Rewrote the `test_init` tests to use the freezer and check the state machine what happens to the entity

joostlek commented:
I think we still need to figure out what the tests in config flow actually test. It now just feels like they're there for touching code, but we're not testing what behaviour we expect it to do

wtaferner commented:
> I'm sick today, but hopefully can finish this up this week

Hope you are healthy again and ready to tackle the last mile 🙏🏼 Really looking forward to have this merged...every local implementation is worth Gold for the Home Assistant community and basically for peace on our beautiful planet 🤣

alistair23 committed:
tests: husqvarna_automower_ble: Remove outdated tests

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 committed:
husqvarna_automower_ble: Update strings

Signed-off-by: Alistair Francis <alistair@alistair23.me>

alistair23 commented:
I have removed the unnecessary tests that weren't really testing anything, rebased on the latest dev branch and updated `strings.json`. I think that's everything addressed

alistair23 commented:
> How are you doing? Let's try to tackle it for this month :)

Good! I think this is good to go, just waiting for the CI

alistair23 commented:
Urgh, it seems like these changes are breaking Gardena: https://github.com/home-assistant/core/actions/runs/11234232316/job/31229988886?pr=108326

joostlek commented:
Can you reproduce it locally? Otherwise it can be a flakey test (I am also not sure if this PR is tested in that bucket)

alistair23 commented:
Locally I get a range of snapshot mismatch failures, but not the same error that the CI shows

alistair23 commented:
Things like this

```
―――――――――――――――――――――――――――――――――――― test_switching ――――――――――――――――――――――――――――――――――――

hass = <HomeAssistant RUNNING>
mock_entry = <ConfigEntry entry_id=01J9NYRRDAJ8DC869S0E6C6VK2 version=1 domain=gardena_bluetooth title=Mock Title state=ConfigEntryState.LOADED unique_id=None>
mock_client = <Mock spec_set='Client' id='139682035100768'>
mock_switch_chars = {'00002a24-0000-1000-8000-00805f9b34fb': b'Mock Model', '00002a26-0000-1000-8000-00805f9b34fb': b'1.2.3', '98bd0f11-0b0e-421a-84e5-ddbf75dc6de4': b'\x00', '98bd0f13-0b0e-421a-84e5-ddbf75dc6de4': b'\x00\x00\x00\x00', ...}

    async def test_switching(
        hass: HomeAssistant,
        mock_entry: MockConfigEntry,
        mock_client: Mock,
        mock_switch_chars: dict[str, bytes],
    ) -> None:
        """Test switching makes correct calls."""
    
        entity_id = "switch.mock_title_open"
        await setup_entry(hass, mock_entry, [Platform.SWITCH])
>       assert hass.states.get(entity_id)
E       AssertionError: assert None
E        +  where None = get('switch.mock_title_open')
E        +    where get = <homeassistant.core.StateMachine object at 0x7f0a4262a520>.get
E        +      where <homeassistant.core.StateMachine object at 0x7f0a4262a520> = <HomeAssistant RUNNING>.states

tests/components/gardena_bluetooth/test_switch.py:66: AssertionError
-------------------------------- Captured stderr setup ------------------
```

alistair23 commented:
@elupus in case you have any ideas about the failure at: https://github.com/home-assistant/core/actions/runs/11234232316/job/31229988886?pr=108326

I think the issue is that Gardena water timers and Husvarna Mowers use the same manufacture (Husqvarna) and service UUIDs. So they conflict and it isn't easy to tell them apart. Any ideas?

alistair23 commented:
Seems to just be a flaky test. Re-ran it and it works now

joostlek committed:
Fix config flow tests

joostlek committed:
Merge branch 'dev' into alistair/husqvarna_automower_ble

joostlek committed:
Fix translation keys

mkmer reviewed (approved):
LGTM. Ready for prime time :)

joostlek commented:
Also as headsup since you said you don't have discord, if you ever feel like using it, feel free to let me know your username, because then you can get access to the integration owners channel where you can read updates, discuss problems and meet others. Just wanted to let you know as headsup :) (no pressure on creating it, it's not a must)

alistair23 commented:
Thanks!

and thanks everyone and especially @joostlek for all of the help!
